core/src/main/scala/org/apache/spark/crypto/CryptoOutputStream.scala

```diff
@@ -0,0 +1,225 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.crypto
+
+import java.io.{IOException, FilterOutputStream, OutputStream}
+import java.nio.ByteBuffer
+import java.security.GeneralSecurityException
+
+import com.google.common.base.Preconditions
+
+import org.apache.spark.Logging
+
+/**
+ * CryptoOutputStream encrypts data. It is not thread-safe. AES CTR mode is
+ * required in order to ensure that the plain text and cipher text have a 1:1
+ * mapping. The encryption is buffer based. The key points of the encryption are
+ * (1) calculating counter and (2) padding through stream position.
+ *
+ * counter = base + pos/(algorithm blocksize);
+ * padding = pos%(algorithm blocksize);
+ *
+ * The underlying stream offset is maintained as state.
+ */
+private[spark] class CryptoOutputStream(
+    out: OutputStream,
+    private[spark] val codec: CryptoCodec,
+    bufferSizeVal: Int,
+    keyVal: Array[Byte],
+    ivVal: Array[Byte],
+    streamOffsetVal: Long,
+    isDirectBuf: Boolean) extends FilterOutputStream(out: OutputStream) with Logging {
+  var encryptor: Encryptor = null
+  var streamOffset: Long = 0
+  /**
+   * Padding = pos%(algorithm blocksize); Padding is put into [[inBuffer]]
+   * before any other data goes in. The purpose of padding is to put input data
+   * at proper position.
+   */
+  var padding: Byte = 0
+  var closed: Boolean = false
+  var key: Array[Byte] = null
+  var initIV: Array[Byte] = null
+  var iv: Array[Byte] = null
+  val oneByteBuf: Array[Byte] = new Array[Byte](1)
+
+  CryptoStreamUtils.checkCodec(codec)
+  var bufferSize = CryptoStreamUtils.checkBufferSize(codec, bufferSizeVal)
+
+  lazy val tmpBuf: Array[Byte] = new Array[Byte](bufferSize)
+
+  key = keyVal.clone()
+  initIV = ivVal.clone()
+  iv = ivVal.clone()
+
+  /**
+   * Input data buffer. The data starts at inBuffer.position() and ends at
+   * inBuffer.limit().
+   */
+  lazy val inBuffer = if (isDirectBuf) {
+    ByteBuffer.allocateDirect(bufferSize)
+  } else {
+    ByteBuffer.allocate(bufferSize)
+  }
+
+  /**
+   * Encrypted data buffer. The data starts at outBuffer.position() and ends at
+   * outBuffer.limit();
+   */
+  lazy val outBuffer = if (isDirectBuf) {
+    ByteBuffer.allocateDirect(bufferSize)
+  } else {
+    ByteBuffer.allocate(bufferSize)
+  }
+  streamOffset = streamOffsetVal
+  encryptor = codec.createEncryptor()
+
+  updateEncryptor()
+
+  def this(
```

Comment:
BTW are all these constructors really needed? If everything goes through `createCryptoOutputStream` and `createCryptoInputStream`, you should only need one constructor in each class.